### DISCUSSION TITLE
Mayby a simple right problem

### DISCUSSION THREAD

**Discussion by ghost** [Rapptz/discord.py#8226] [General] [answered]
**Hi there**

I try to learn more from python while using a discord bot. Everything works fine but im trying new stuff to learn.

What I try this time is

1. a command, where the bot creates a role if not exist and gets the user to this role
2. make a punch of roles in a row 

1)
A User should use a command to get the right role for him. At this point @everyone has no right to access roles.
But the bot does, so isnt it enouht that the bot have the rights? Because if I try it, this happens:
as noone using this command and the role exist, I join the role
as noone using the command and role dont exist, I get an error, not makin the new role
as admin it works

My try looks like:

<pre>
@client.command(aliases=['location','loc'])
async def newlocation(ctx, role: str=None):
	user=ctx.message.author
	if role is None:
		await ctx.send(f"{user.mention}```You have to choose a location\nBsp.: !location USA\n```")
	else:
		if role!='Administrator':
			role=discord.utils.get(ctx.guild.roles,name=role)
			if role is None:
				await ctx.guild.create_role(name=role)
			if role not in user.roles:
				await discord.Member.add_roles(user,role)
				await ctx.send(f"{user.mention}```Your location is saved```")
			else:
				await ctx.send(f"{user.mention}```You are still in this group```")
</pre>
So is it only because default user has not the right?

2)
I wanted to get faster input make a command to set up multiple roles.  I tried at first time with 2 or 3 roles to see if it works. 
My code seems to work, it puts the 3 new roles to my roles. Then I only changed the values of the arrey and now nothing works, but I also get no error output. Nothing happens and no new role is in my list.

My try looks like:
<pre>
@client.command(aliases=['add'])
async def addnew(ctx):
	addroles=['USA','CANADA',...... in my example i think i have about 90 entries]
	for role in addroles:
		await ctx.guild.create_role(name=role)
		await ctx.send(f"{role} created")
</pre>

After I deletet all new roles and add the over 80 in the arrey, the code dont make new roles, is I put the send code befor the create.role one, I get the mess but not the role and no error.

Can someone help me looking throught the trees?
Thanks for your time :-)

**Comment by LostLuma** [comment, marked as answer]
Hello!

Your first issue is caused by re-using the same variable multiple times in an unlucky configuration. You are correct that only the bot needs permissions to create and assign roles, rather than users using the commands. Let's step through the relevant part of the code:

```py
@client.command(aliases=['location', 'loc'])
async def newlocation(ctx, role: str = None):
    ...

    # If no role matching the content of the `role` variable exists the variable will be set to `None` here
    role = discord.utils.get(ctx.guild.roles, name=role)

    if role is None:
        # A new role is being created, however the `name` argument being passed will always be `None`
        # As the `role` variable is `None`, meaning every role being created here will be called "new role"
        await ctx.guild.create_role(name=role)

    # At this point the `role` variable is still `None`, it will not be a part of `user.roles`
    if role not in user.roles:
        # This call should use the Member instance, not the class
        # You might want to read up on classes and instances elsewhere
        await discord.Member.add_roles(user,role)
        await ctx.send(f"{user.mention}```Your location is saved```")
```

I would suggest naming the argument for the user's location in your command something like `location` to avoid this ambiguity.

```py
@client.command(aliases=['location', 'loc'])
async def newlocation(ctx, location: str = None):
```

Secondly, if you'd like to assign the newly created role to a user you will need the actual [`Role`](https://discordpy.readthedocs.io/en/stable/api.html#discord.Role) object instance. You have already done this using `utils.get` on existing roles, however note that running [`create_role`](https://discordpy.readthedocs.io/en/stable/api.html#discord.Guild.create_role) will also return the role and can be used:

```py
role = await ctx.guild.create_role(name=location)
await ctx.author.add_roles(role)
```

Your second issue is likely because there are additional anti abuse rate-limits on the API which could for example limit the amount of roles you can create within a single day, rather than an issue with your code.

I would suggest waiting this one out, and also not running this sort of operation very often or at all in the future to avoid running into it again. Creating roles as-needed with the `newlocation` command only for example is a rather elegant solution.

In the future please also post the full error you receive when asking for help, it makes everyone else's job a lot easier.

**Reply by ghost** [reply]
Wow wow wow wow. Tousend of thanks for you. this kind of detailed answer is really great. Thanks a lot.
I will try to fix it later and repost what my hopefully GOOD solution will be.

Thumbs up for @LostLuma

**Comment by ghost** [comment]
one another question, I didnt try this one yet, but if I do like:

<pre>
@client.command(aliases=['s', 'standort', 'Standort', 'ort', 'Ort'])
async def newlocation(ctx, location: str=None):
	if location is None:
		await ctx.send(f'{ctx.message.author.mention}```you need to use a location```')
	else:
		role = await ctx.guild.create_role(name=location)
		await ctx.author.add_roles(role)
</pre>

will he add the role even if it exist? do i need a if else for the role?

**Reply by LostLuma** [reply]
You will want to check whether a role with the name already exists else it will create a second role, yes.

**Reply by ghost** [reply]
ok thx

**Comment by ghost** [comment]
So, I can try it at evening, not now, but this is my solution. Waht do you think?

<pre>
@client.command(aliases=['location'])
async def newlocation(ctx, location: str=None):
	if location is None:
		await ctx.send(f'{ctx.message.author.mention}```Location is empty, I dont think you live nowhere :)```')
		return False
	else:
		role = discord.utils.get(ctx.guild.roles, name=location)
		if not role:
			role = await ctx.guild.create_role(name=location)
			await ctx.send(f"{ctx.message.author.mention}```Creating location first```")

		if role not in ctx.message.author.roles:
			await ctx.send(f"{ctx.message.author.mention}```Your location has been saved```")
			await ctx.author.add_roles(role)
		else:
			await ctx.send(f"{ctx.message.author.mention}```You allready saved your location```")
</pre>

**Reply by LostLuma** [reply]
This seems fine!

**Reply by ghost** [reply]
I thank you a lot :-)

Is there a possibility to get banned or something like this as a bot id from discord for the limit thing? I tried the code wit a dummy server and it worked, but at mine it does not, i think it happened at the command with over 80 roles to be created and that i tried as a newby to often for getting the right code, so mayby my bot is out of creation permission now :-(

**Reply by LostLuma** [reply]
You can get banned, however if you did you would either receive a special error or a system DM from Discord informing you about it (depends on the endpoint). If it just seems to be stuck waiting you just need to wait.

**Reply by ghost** [reply]
sounds good, i had no message so only need to wait, thx
